lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -350,3 +352,185 @@ def _get(self, json_data: JSON, _) -> Response:
+
+            dataset: The associated Dataset object
+        """
+        raise NotImplementedError()
+
+    def complete(self, dataset: Dataset, json_data: JSON, error_count: int) -> None:
+        """
+        Complete a bulk Elasticsearch operation, generally by modifying the
+        source Dataset resource.
+
+        This is an abstract method that must be implemented by a subclass.
+
+        Args:
+
+            dataset: The associated Dataset object
+
+            json_data: JSON dictionary of type-normalized key-value pairs
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset
+
+            error_count: The number of errors encountered during the bulk
+                operation. We want to support idempotency: the Dataset should
+                not be altered unless the error_count is 0.
+        """
+        raise NotImplementedError()
+
+    def _post(self, json_data: JSON, _) -> Response:
+        """
+        Perform the requested POST operation, and handle any exceptions.
+
+        NOTE: This is called by the ApiBase post() method through its dispatch
+        method, which provides parameter validation.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            _: Original incoming Request object (not used)
+
+        Returns:
+            Response to return to client
+        """
+        klasname = self.__class__.__name__
+
+        try:
+            dataset = Dataset.attach(
+                controller=json_data["controller"], name=json_data["name"]
+            )
+        except DatasetNotFound as e:
+            abort(HTTPStatus.NOT_FOUND, message=str(e))
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For bulk Elasticsearch operations, we check authorization against the
+        # ownership of a designated dataset rather than having an explicit
+        # "user" JSON parameter. This will raise UnauthorizedAccess on failure.
+        try:
+            self._check_authorization(owner.username, json_data["access"])
+        except UnauthorizedAccess as e:
+            abort(HTTPStatus.FORBIDDEN, message=str(e))
+
+        # Build an Elasticsearch instance to manage the bulk update
+        elastic = elasticsearch.Elasticsearch([self.es_url])
+
+        try:
+            # Pass the bulk command generator to the helper
+            results = elasticsearch.helpers.streaming_bulk(
+                elastic,
+                self.generate_documents(json_data, dataset),
+                raise_on_exception=False,
+                raise_on_error=False,
+            )
+        except Exception as e:
```

Comment:
Keep in mind that line 477 returns a generator object.  If an exception is raised while creating the generator object, this try catch will be useful.  But that is why we have `py-es-bulk`.  =)  The exception handling needs to wrap the `for` loop at line 494 below.